File: android/app/src/main/java/com/newcam/CCCameraManager.java
package com.newcam;

import android.content.Intent;

import com.facebook.react.common.MapBuilder;
import com.facebook.react.uimanager.ViewGroupManager;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.annotations.ReactProp;

import java.util.Map;

import javax.annotation.Nullable;

/**
 * Created by dan on 12/16/16.
 */

public class CCCameraManager extends ViewGroupManager<CCCameraView> {

    public static final String REACT_CLASS = "CompanyCamCamera";

    private static CCCameraView latestView;

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    @Override
    protected CCCameraView createViewInstance(ThemedReactContext context) {
        latestView = new CCCameraView(context);
        return latestView;
    }

    public static CCCameraView getLatestView(){
        return latestView;
    }

    @ReactProp(name = "storagePath")
    public void setStoragePath(CCCameraView view, String str){
        view.setStoragePath(str);
    }

    @ReactProp(name = "projectName")
    public void setProjectName(CCCameraView view, String str){
        view.setProjectName(str);
    }

    @ReactProp(name = "projectAddress")
    public void setProjectAddress(CCCameraView view, String str){
        view.setProjectAddress(str);
    }

    @ReactProp(name = "exifLat")
    public void setExifLat(CCCameraView view, double val){
        view.setExifLat(val);
    }

    @ReactProp(name = "exifLon")
    public void setExifLon(CCCameraView view, double val){
        view.setExifLon(val);
    }

    @ReactProp(name = "exifLocTimestamp")
    public void setExifLocTimestamp(CCCameraView view, double val){
        view.setExifLocTimestamp(val);
    }

    @ReactProp(name = "auxModeCaption")
    public void setAuxModeCaption(CCCameraView view, String val){
        view.setAuxModeCaption(val);
    }

    @ReactProp(name = "hideNativeUI")
    public void setHideNativeUI(CCCameraView view, boolean val){
        view.setHideNativeUI(val);
    }

    @Override
    @Nullable
    public Map getExportedCustomDirectEventTypeConstants() {
        return MapBuilder.of(
            "onClose",
            MapBuilder.of("registrationName", "onClose"),
            "onPhotoAccepted",
            MapBuilder.of("registrationName", "onPhotoAccepted"),
            "onPhotoTaken",
            MapBuilder.of("registrationName", "onPhotoTaken"),
            "onAuxModeClicked",
            MapBuilder.of("registrationName", "onAuxModeClicked")
        );
    }
}

File: ios/CCCameraManager.m
//
//  CCCameraManager.m
//  newcam
//
//  Created by Matt Boyd on 5/10/17.
//  Copyright © 2017 Facebook. All rights reserved.
//

#import "CCCameraManager.h"

@implementation CCCameraManager

static CCCameraView *latestView;

RCT_EXPORT_MODULE(CompanyCamCamera)

-(UIView *)view {

    if (latestView == nil) {
        latestView = [[CCCameraView alloc] initWithManager:self bridge:self.bridge];
    }
    else {
        [latestView setupView];
    }

    // Create the CCCamera object
    latestView.camera = [[CCCamera alloc] init];

    // Set the layout object's reference to the camera
    [latestView.cameraLayout setCameraObject:latestView.camera];

    return latestView;
}

+(CCCameraView *)getLatestView {
    return latestView;
}

#pragma mark React props

RCT_CUSTOM_VIEW_PROPERTY(storagePath, NSString, CCCameraView) {
    NSString *storagePath = [RCTConvert NSString:json];
    [latestView setStoragePath:storagePath];
}

RCT_CUSTOM_VIEW_PROPERTY(projectName, NSString, CCCameraView) {
    NSString *projectName = [RCTConvert NSString:json];
    [latestView setProjectName:projectName];
}

RCT_CUSTOM_VIEW_PROPERTY(projectAddress, NSString, CCCameraView) {
    NSString *projectAddress = [RCTConvert NSString:json];
    [latestView setProjectAddress:projectAddress];
}

RCT_CUSTOM_VIEW_PROPERTY(exifLat, double, CCCameraView) {
    double exifLat = [RCTConvert double:json];
    [latestView setExifLat:exifLat];
}

RCT_CUSTOM_VIEW_PROPERTY(exifLon, double, CCCameraView) {
    double exifLon = [RCTConvert double:json];
    [latestView setExifLon:exifLon];
}

RCT_CUSTOM_VIEW_PROPERTY(exifLocTimestamp, double, CCCameraView) {
    double exifLocTimestamp = [RCTConvert double:json];
    [latestView setExifLocTimestamp:exifLocTimestamp];
}

RCT_CUSTOM_VIEW_PROPERTY(hideNativeUI, BOOL, CCCameraView) {
    BOOL hideNativeUI = [RCTConvert BOOL:json];
    [latestView setHideNativeUI:hideNativeUI];
}

RCT_EXPORT_VIEW_PROPERTY(onClose, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onPhotoTaken, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onPhotoAccepted, RCTDirectEventBlock)

RCT_CUSTOM_VIEW_PROPERTY(flashMode, NSInteger, CCCameraFlashMode) {
    CCCameraFlashMode mode = [RCTConvert NSInteger:json];
    [self setFlashMode:mode];
}

- (void)setFlashMode:(CCCameraFlashMode)flashMode {
    AVCaptureDevice *device = [[latestView.camera deviceInput] device];
    NSError *error = nil;

    AVCaptureTorchMode torchMode;

    // Limit to torch on/off
    switch (flashMode) {
        case CCCameraFlashModeAuto:
            torchMode = AVCaptureTorchModeOff;
            break;
        case CCCameraFlashModeOn:
            torchMode = AVCaptureTorchModeOff;
            break;
        case CCCameraFlashModeOff:
            torchMode = AVCaptureTorchModeOff;
            break;
        case CCCameraFlashModeTorch:
            torchMode = AVCaptureTorchModeOn;
            break;
        default:
            torchMode = AVCaptureTorchModeOff;
            break;
    }

    if (![device lockForConfiguration:&error]) {
        NSLog(@"%@", error);
        return;
    }
    if (device.hasTorch && [device isTorchModeSupported:torchMode])
    {
        NSError *error = nil;
        if ([device lockForConfiguration:&error])
        {
            [device setTorchMode:torchMode];
            [device unlockForConfiguration];
        }
        else
        {
            NSLog(@"%@", error);
        }
    }
    [device unlockForConfiguration];
}

RCT_CUSTOM_VIEW_PROPERTY(cameraMode, NSInteger, CCCameraMode) {
    CCCameraMode mode = [RCTConvert NSInteger:json];
    [latestView.camera changeCameraMode:mode];
}

RCT_CUSTOM_VIEW_PROPERTY(resolutionMode, NSInteger, CCCameraResolutionMode) {
    CCCameraResolutionMode mode = [RCTConvert NSInteger:json];
    [latestView.camera changeResolution:mode];
}


@end
